// modern-treasury/resources/payment-orders/payment-orders.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as ReversalsAPI from "./reversals.mjs";
import { Reversals, } from "./reversals.mjs";
import { Page } from "../../core/pagination.mjs";
import { maybeMultipartFormRequestOptions } from "../../internal/uploads.mjs";
import { path } from "../../internal/utils/path.mjs";
export class PaymentOrders extends APIResource {
    constructor() {
        super(...arguments);
        this.reversals = new ReversalsAPI.Reversals(this._client);
    }
    /**
     * Create a new Payment Order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.create({
     *   amount: 0,
     *   direction: 'credit',
     *   originating_account_id:
     *     '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   type: 'ach',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/payment_orders', maybeMultipartFormRequestOptions({ body, ...options }, this._client));
    }
    /**
     * Get details on a single payment order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/payment_orders/${id}`, options);
    }
    /**
     * Update a payment order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.update(
     *   'id',
     * );
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/payment_orders/${id}`, { body, ...options });
    }
    /**
     * Get a list of all payment orders
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const paymentOrder of client.paymentOrders.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/payment_orders', (Page), { query, ...options });
    }
    /**
     * Create a new payment order asynchronously
     *
     * @example
     * ```ts
     * const asyncResponse =
     *   await client.paymentOrders.createAsync({
     *     amount: 0,
     *     direction: 'credit',
     *     originating_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     type: 'ach',
     *   });
     * ```
     */
    createAsync(body, options) {
        return this._client.post('/api/payment_orders/create_async', { body, ...options });
    }
}
PaymentOrders.Reversals = Reversals;
