// modern-treasury/resources/payment-orders/payment-orders.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentOrders = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const ReversalsAPI = tslib_1.__importStar(require("./reversals.js"));
const reversals_1 = require("./reversals.js");
const pagination_1 = require("../../core/pagination.js");
const uploads_1 = require("../../internal/uploads.js");
const path_1 = require("../../internal/utils/path.js");
class PaymentOrders extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.reversals = new ReversalsAPI.Reversals(this._client);
    }
    /**
     * Create a new Payment Order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.create({
     *   amount: 0,
     *   direction: 'credit',
     *   originating_account_id:
     *     '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   type: 'ach',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/payment_orders', (0, uploads_1.maybeMultipartFormRequestOptions)({ body, ...options }, this._client));
    }
    /**
     * Get details on a single payment order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/payment_orders/${id}`, options);
    }
    /**
     * Update a payment order
     *
     * @example
     * ```ts
     * const paymentOrder = await client.paymentOrders.update(
     *   'id',
     * );
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/payment_orders/${id}`, { body, ...options });
    }
    /**
     * Get a list of all payment orders
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const paymentOrder of client.paymentOrders.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/payment_orders', (pagination_1.Page), { query, ...options });
    }
    /**
     * Create a new payment order asynchronously
     *
     * @example
     * ```ts
     * const asyncResponse =
     *   await client.paymentOrders.createAsync({
     *     amount: 0,
     *     direction: 'credit',
     *     originating_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     type: 'ach',
     *   });
     * ```
     */
    createAsync(body, options) {
        return this._client.post('/api/payment_orders/create_async', { body, ...options });
    }
}
exports.PaymentOrders = PaymentOrders;
PaymentOrders.Reversals = reversals_1.Reversals;
