// modern-treasury/resources/internal-accounts/balance-reports.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { Page } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class BalanceReports extends APIResource {
    /**
     * create balance reports
     *
     * @example
     * ```ts
     * const balanceReport =
     *   await client.internalAccounts.balanceReports.create(
     *     'internal_account_id',
     *     {
     *       as_of_date: '2019-12-27',
     *       as_of_time: 'as_of_time',
     *       balance_report_type: 'intraday',
     *       balances: [
     *         {
     *           amount: 0,
     *           balance_type: 'closing_available',
     *           vendor_code: 'vendor_code',
     *           vendor_code_type: 'vendor_code_type',
     *         },
     *       ],
     *     },
     *   );
     * ```
     */
    create(internalAccountID, body, options) {
        return this._client.post(path `/api/internal_accounts/${internalAccountID}/balance_reports`, {
            body,
            ...options,
        });
    }
    /**
     * Get a single balance report for a given internal account.
     *
     * @example
     * ```ts
     * const balanceReport =
     *   await client.internalAccounts.balanceReports.retrieve(
     *     'string',
     *     { internal_account_id: 'internal_account_id' },
     *   );
     * ```
     */
    retrieve(id, params, options) {
        const { internal_account_id } = params;
        return this._client.get(path `/api/internal_accounts/${internal_account_id}/balance_reports/${id}`, options);
    }
    /**
     * Get all balance reports for a given internal account.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const balanceReport of client.internalAccounts.balanceReports.list(
     *   'internal_account_id',
     * )) {
     *   // ...
     * }
     * ```
     */
    list(internalAccountID, query = {}, options) {
        return this._client.getAPIList(path `/api/internal_accounts/${internalAccountID}/balance_reports`, (Page), { query, ...options });
    }
    /**
     * Deletes a given balance report.
     *
     * @example
     * ```ts
     * await client.internalAccounts.balanceReports.delete('id', {
     *   internal_account_id: 'internal_account_id',
     * });
     * ```
     */
    delete(id, params, options) {
        const { internal_account_id } = params;
        return this._client.delete(path `/api/internal_accounts/${internal_account_id}/balance_reports/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
