// modern-treasury/resources/internal-accounts/internal-accounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as BalanceReportsAPI from "./balance-reports.mjs";
import { BalanceReports, } from "./balance-reports.mjs";
import { Page } from "../../core/pagination.mjs";
import { path } from "../../internal/utils/path.mjs";
export class InternalAccounts extends APIResource {
    constructor() {
        super(...arguments);
        this.balanceReports = new BalanceReportsAPI.BalanceReports(this._client);
    }
    /**
     * create internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.create({
     *     connection_id: 'connection_id',
     *     currency: 'USD',
     *     name: 'name',
     *     party_name: 'party_name',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/internal_accounts', { body, ...options });
    }
    /**
     * get internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/internal_accounts/${id}`, options);
    }
    /**
     * update internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/internal_accounts/${id}`, { body, ...options });
    }
    /**
     * list internal accounts
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const internalAccount of client.internalAccounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/internal_accounts', (Page), { query, ...options });
    }
    /**
     * update account_capability
     *
     * @example
     * ```ts
     * const response =
     *   await client.internalAccounts.updateAccountCapability(
     *     'id',
     *     {
     *       internal_account_id: 'internal_account_id',
     *       identifier: 'identifier',
     *     },
     *   );
     * ```
     */
    updateAccountCapability(id, params, options) {
        const { internal_account_id, ...body } = params;
        return this._client.patch(path `/api/internal_accounts/${internal_account_id}/account_capabilities/${id}`, { body, ...options });
    }
}
InternalAccounts.BalanceReports = BalanceReports;
