// modern-treasury/resources/internal-accounts/internal-accounts.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalAccounts = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const BalanceReportsAPI = tslib_1.__importStar(require("./balance-reports.js"));
const balance_reports_1 = require("./balance-reports.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class InternalAccounts extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.balanceReports = new BalanceReportsAPI.BalanceReports(this._client);
    }
    /**
     * create internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.create({
     *     connection_id: 'connection_id',
     *     currency: 'USD',
     *     name: 'name',
     *     party_name: 'party_name',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/internal_accounts', { body, ...options });
    }
    /**
     * get internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/internal_accounts/${id}`, options);
    }
    /**
     * update internal account
     *
     * @example
     * ```ts
     * const internalAccount =
     *   await client.internalAccounts.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/internal_accounts/${id}`, { body, ...options });
    }
    /**
     * list internal accounts
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const internalAccount of client.internalAccounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/internal_accounts', (pagination_1.Page), { query, ...options });
    }
    /**
     * update account_capability
     *
     * @example
     * ```ts
     * const response =
     *   await client.internalAccounts.updateAccountCapability(
     *     'id',
     *     {
     *       internal_account_id: 'internal_account_id',
     *       identifier: 'identifier',
     *     },
     *   );
     * ```
     */
    updateAccountCapability(id, params, options) {
        const { internal_account_id, ...body } = params;
        return this._client.patch((0, path_1.path) `/api/internal_accounts/${internal_account_id}/account_capabilities/${id}`, { body, ...options });
    }
}
exports.InternalAccounts = InternalAccounts;
InternalAccounts.BalanceReports = balance_reports_1.BalanceReports;
