// @datadog/browser-rum-react/esm/domain/reactPlugin.js
var globalPublicApi;
var globalConfiguration;
var onReactPluginInitSubscribers = [];
export function reactPlugin(configuration) {
    if (configuration === void 0) { configuration = {}; }
    return {
        name: 'react',
        onInit: function (_a) {
            var publicApi = _a.publicApi, initConfiguration = _a.initConfiguration;
            globalPublicApi = publicApi;
            globalConfiguration = configuration;
            onReactPluginInitSubscribers.forEach(function (subscriber) { return subscriber(configuration, publicApi); });
            if (configuration.router) {
                initConfiguration.trackViewsManually = true;
            }
        },
        getConfigurationTelemetry: function () {
            return { router: !!configuration.router };
        },
    };
}
export function onReactPluginInit(callback) {
    if (globalConfiguration && globalPublicApi) {
        callback(globalConfiguration, globalPublicApi);
    }
    else {
        onReactPluginInitSubscribers.push(callback);
    }
}
export function resetReactPlugin() {
    globalPublicApi = undefined;
    globalConfiguration = undefined;
    onReactPluginInitSubscribers.length = 0;
}
