// modern-treasury/resources/account-details.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountDetails = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class AccountDetails extends resource_1.APIResource {
    /**
     * Create an account detail for an external account.
     */
    create(accountID, params, options) {
        const { accounts_type, ...body } = params;
        return this._client.post((0, path_1.path) `/api/${accounts_type}/${accountID}/account_details`, { body, ...options });
    }
    /**
     * Get a single account detail for a single internal or external account.
     */
    retrieve(id, params, options) {
        const { accounts_type, account_id } = params;
        return this._client.get((0, path_1.path) `/api/${accounts_type}/${account_id}/account_details/${id}`, options);
    }
    /**
     * Get a list of account details for a single internal or external account.
     */
    list(accountID, params, options) {
        const { accounts_type, ...query } = params;
        return this._client.getAPIList((0, path_1.path) `/api/${accounts_type}/${accountID}/account_details`, (pagination_1.Page), { query, ...options });
    }
    /**
     * Delete a single account detail for an external account.
     */
    delete(id, params, options) {
        const { accounts_type, account_id } = params;
        return this._client.delete((0, path_1.path) `/api/${accounts_type}/${account_id}/account_details/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.AccountDetails = AccountDetails;
