// modern-treasury/resources/account-details.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { buildHeaders } from "../internal/headers.mjs";
import { path } from "../internal/utils/path.mjs";
export class AccountDetails extends APIResource {
    /**
     * Create an account detail for an external account.
     */
    create(accountID, params, options) {
        const { accounts_type, ...body } = params;
        return this._client.post(path `/api/${accounts_type}/${accountID}/account_details`, { body, ...options });
    }
    /**
     * Get a single account detail for a single internal or external account.
     */
    retrieve(id, params, options) {
        const { accounts_type, account_id } = params;
        return this._client.get(path `/api/${accounts_type}/${account_id}/account_details/${id}`, options);
    }
    /**
     * Get a list of account details for a single internal or external account.
     */
    list(accountID, params, options) {
        const { accounts_type, ...query } = params;
        return this._client.getAPIList(path `/api/${accounts_type}/${accountID}/account_details`, (Page), { query, ...options });
    }
    /**
     * Delete a single account detail for an external account.
     */
    delete(id, params, options) {
        const { accounts_type, account_id } = params;
        return this._client.delete(path `/api/${accounts_type}/${account_id}/account_details/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
