// modern-treasury/resources/counterparties.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Counterparties = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class Counterparties extends resource_1.APIResource {
    /**
     * Create a new counterparty.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.create({
     *   name: 'name',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/counterparties', { body, ...options });
    }
    /**
     * Get details on a single counterparty.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/counterparties/${id}`, options);
    }
    /**
     * Updates a given counterparty with new information.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.update(
     *   'id',
     * );
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/counterparties/${id}`, { body, ...options });
    }
    /**
     * Get a paginated list of all counterparties.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const counterparty of client.counterparties.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/counterparties', (pagination_1.Page), { query, ...options });
    }
    /**
     * Deletes a given counterparty.
     *
     * @example
     * ```ts
     * await client.counterparties.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/counterparties/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Send an email requesting account details.
     *
     * @example
     * ```ts
     * const counterpartyCollectAccountResponse =
     *   await client.counterparties.collectAccount('id', {
     *     direction: 'credit',
     *   });
     * ```
     */
    collectAccount(id, body, options) {
        return this._client.post((0, path_1.path) `/api/counterparties/${id}/collect_account`, { body, ...options });
    }
}
exports.Counterparties = Counterparties;
