// modern-treasury/resources/counterparties.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { buildHeaders } from "../internal/headers.mjs";
import { path } from "../internal/utils/path.mjs";
export class Counterparties extends APIResource {
    /**
     * Create a new counterparty.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.create({
     *   name: 'name',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/counterparties', { body, ...options });
    }
    /**
     * Get details on a single counterparty.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/counterparties/${id}`, options);
    }
    /**
     * Updates a given counterparty with new information.
     *
     * @example
     * ```ts
     * const counterparty = await client.counterparties.update(
     *   'id',
     * );
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/counterparties/${id}`, { body, ...options });
    }
    /**
     * Get a paginated list of all counterparties.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const counterparty of client.counterparties.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/counterparties', (Page), { query, ...options });
    }
    /**
     * Deletes a given counterparty.
     *
     * @example
     * ```ts
     * await client.counterparties.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/counterparties/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Send an email requesting account details.
     *
     * @example
     * ```ts
     * const counterpartyCollectAccountResponse =
     *   await client.counterparties.collectAccount('id', {
     *     direction: 'credit',
     *   });
     * ```
     */
    collectAccount(id, body, options) {
        return this._client.post(path `/api/counterparties/${id}/collect_account`, { body, ...options });
    }
}
