// modern-treasury/resources/expected-payments.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class ExpectedPayments extends APIResource {
    /**
     * create expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.create();
     * ```
     */
    create(body = {}, options) {
        return this._client.post('/api/expected_payments', { body, ...options });
    }
    /**
     * get expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/expected_payments/${id}`, options);
    }
    /**
     * update expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/expected_payments/${id}`, { body, ...options });
    }
    /**
     * list expected_payments
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const expectedPayment of client.expectedPayments.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/expected_payments', (Page), { query, ...options });
    }
    /**
     * delete expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/expected_payments/${id}`, options);
    }
}
