// modern-treasury/resources/expected-payments.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpectedPayments = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class ExpectedPayments extends resource_1.APIResource {
    /**
     * create expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.create();
     * ```
     */
    create(body = {}, options) {
        return this._client.post('/api/expected_payments', { body, ...options });
    }
    /**
     * get expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/expected_payments/${id}`, options);
    }
    /**
     * update expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/expected_payments/${id}`, { body, ...options });
    }
    /**
     * list expected_payments
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const expectedPayment of client.expectedPayments.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/expected_payments', (pagination_1.Page), { query, ...options });
    }
    /**
     * delete expected payment
     *
     * @example
     * ```ts
     * const expectedPayment =
     *   await client.expectedPayments.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/expected_payments/${id}`, options);
    }
}
exports.ExpectedPayments = ExpectedPayments;
