// modern-treasury/resources/external-accounts.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalAccounts = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class ExternalAccounts extends resource_1.APIResource {
    /**
     * create external account
     *
     * @example
     * ```ts
     * const externalAccount =
     *   await client.externalAccounts.create({
     *     counterparty_id: '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/external_accounts', { body, ...options });
    }
    /**
     * show external account
     *
     * @example
     * ```ts
     * const externalAccount =
     *   await client.externalAccounts.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/external_accounts/${id}`, options);
    }
    /**
     * update external account
     *
     * @example
     * ```ts
     * const externalAccount =
     *   await client.externalAccounts.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/external_accounts/${id}`, { body, ...options });
    }
    /**
     * list external accounts
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const externalAccount of client.externalAccounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/external_accounts', (pagination_1.Page), { query, ...options });
    }
    /**
     * delete external account
     *
     * @example
     * ```ts
     * await client.externalAccounts.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/external_accounts/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * complete verification of external account
     *
     * @example
     * ```ts
     * const externalAccount =
     *   await client.externalAccounts.completeVerification('id');
     * ```
     */
    completeVerification(id, body = {}, options) {
        return this._client.post((0, path_1.path) `/api/external_accounts/${id}/complete_verification`, { body, ...options });
    }
    /**
     * verify external account
     *
     * @example
     * ```ts
     * const response = await client.externalAccounts.verify(
     *   'id',
     *   {
     *     originating_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     payment_type: 'ach',
     *   },
     * );
     * ```
     */
    verify(id, body, options) {
        return this._client.post((0, path_1.path) `/api/external_accounts/${id}/verify`, { body, ...options });
    }
}
exports.ExternalAccounts = ExternalAccounts;
