// modern-treasury/resources/ledger-account-balance-monitors.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerAccountBalanceMonitors = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class LedgerAccountBalanceMonitors extends resource_1.APIResource {
    /**
     * Create a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.create({
     *     alert_condition: {
     *       field: 'field',
     *       operator: 'operator',
     *       value: 0,
     *     },
     *     ledger_account_id: 'ledger_account_id',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_account_balance_monitors', { body, ...options });
    }
    /**
     * Get details on a single ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/ledger_account_balance_monitors/${id}`, options);
    }
    /**
     * Update a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/ledger_account_balance_monitors/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger account balance monitors.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccountBalanceMonitor of client.ledgerAccountBalanceMonitors.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_account_balance_monitors', (pagination_1.Page), { query, ...options });
    }
    /**
     * Delete a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/ledger_account_balance_monitors/${id}`, options);
    }
}
exports.LedgerAccountBalanceMonitors = LedgerAccountBalanceMonitors;
