// modern-treasury/resources/ledger-account-balance-monitors.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class LedgerAccountBalanceMonitors extends APIResource {
    /**
     * Create a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.create({
     *     alert_condition: {
     *       field: 'field',
     *       operator: 'operator',
     *       value: 0,
     *     },
     *     ledger_account_id: 'ledger_account_id',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_account_balance_monitors', { body, ...options });
    }
    /**
     * Get details on a single ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/ledger_account_balance_monitors/${id}`, options);
    }
    /**
     * Update a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/ledger_account_balance_monitors/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger account balance monitors.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccountBalanceMonitor of client.ledgerAccountBalanceMonitors.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_account_balance_monitors', (Page), { query, ...options });
    }
    /**
     * Delete a ledger account balance monitor.
     *
     * @example
     * ```ts
     * const ledgerAccountBalanceMonitor =
     *   await client.ledgerAccountBalanceMonitors.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/ledger_account_balance_monitors/${id}`, options);
    }
}
