// modern-treasury/resources/ledgers.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ledgers = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class Ledgers extends resource_1.APIResource {
    /**
     * Create a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.create({
     *   name: 'name',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledgers', { body, ...options });
    }
    /**
     * Get details on a single ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/ledgers/${id}`, options);
    }
    /**
     * Update the details of a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/ledgers/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledgers.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledger of client.ledgers.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledgers', (pagination_1.Page), { query, ...options });
    }
    /**
     * Delete a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/ledgers/${id}`, options);
    }
}
exports.Ledgers = Ledgers;
