// modern-treasury/resources/ledgers.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class Ledgers extends APIResource {
    /**
     * Create a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.create({
     *   name: 'name',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledgers', { body, ...options });
    }
    /**
     * Get details on a single ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/ledgers/${id}`, options);
    }
    /**
     * Update the details of a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/ledgers/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledgers.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledger of client.ledgers.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledgers', (Page), { query, ...options });
    }
    /**
     * Delete a ledger.
     *
     * @example
     * ```ts
     * const ledger = await client.ledgers.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/ledgers/${id}`, options);
    }
}
