// modern-treasury/resources/ledger-accounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class LedgerAccounts extends APIResource {
    /**
     * Create a ledger account.
     *
     * @example
     * ```ts
     * const ledgerAccount = await client.ledgerAccounts.create({
     *   currency: 'currency',
     *   ledger_id: '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   name: 'name',
     *   normal_balance: 'credit',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_accounts', { body, ...options });
    }
    /**
     * Get details on a single ledger account.
     *
     * @example
     * ```ts
     * const ledgerAccount = await client.ledgerAccounts.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, query = {}, options) {
        return this._client.get(path `/api/ledger_accounts/${id}`, { query, ...options });
    }
    /**
     * Update the details of a ledger account.
     *
     * @example
     * ```ts
     * const ledgerAccount = await client.ledgerAccounts.update(
     *   'id',
     * );
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/ledger_accounts/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger accounts.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccount of client.ledgerAccounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_accounts', (Page), { query, ...options });
    }
    /**
     * Delete a ledger account.
     *
     * @example
     * ```ts
     * const ledgerAccount = await client.ledgerAccounts.delete(
     *   'id',
     * );
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/ledger_accounts/${id}`, options);
    }
}
