// modern-treasury/resources/ledger-account-categories.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerAccountCategories = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class LedgerAccountCategories extends resource_1.APIResource {
    /**
     * Create a ledger account category.
     *
     * @example
     * ```ts
     * const ledgerAccountCategory =
     *   await client.ledgerAccountCategories.create({
     *     currency: 'currency',
     *     ledger_id: '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     name: 'name',
     *     normal_balance: 'credit',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_account_categories', { body, ...options });
    }
    /**
     * Get the details on a single ledger account category.
     *
     * @example
     * ```ts
     * const ledgerAccountCategory =
     *   await client.ledgerAccountCategories.retrieve('id');
     * ```
     */
    retrieve(id, query = {}, options) {
        return this._client.get((0, path_1.path) `/api/ledger_account_categories/${id}`, { query, ...options });
    }
    /**
     * Update the details of a ledger account category.
     *
     * @example
     * ```ts
     * const ledgerAccountCategory =
     *   await client.ledgerAccountCategories.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/ledger_account_categories/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger account categories.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccountCategory of client.ledgerAccountCategories.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_account_categories', (pagination_1.Page), {
            query,
            ...options,
        });
    }
    /**
     * Delete a ledger account category.
     *
     * @example
     * ```ts
     * const ledgerAccountCategory =
     *   await client.ledgerAccountCategories.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/ledger_account_categories/${id}`, options);
    }
    /**
     * Add a ledger account to a ledger account category.
     *
     * @example
     * ```ts
     * await client.ledgerAccountCategories.addLedgerAccount(
     *   'ledger_account_id',
     *   { id: 'id' },
     * );
     * ```
     */
    addLedgerAccount(ledgerAccountID, params, options) {
        const { id } = params;
        return this._client.put((0, path_1.path) `/api/ledger_account_categories/${id}/ledger_accounts/${ledgerAccountID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Add a ledger account category to a ledger account category.
     *
     * @example
     * ```ts
     * await client.ledgerAccountCategories.addNestedCategory(
     *   'sub_category_id',
     *   { id: 'id' },
     * );
     * ```
     */
    addNestedCategory(subCategoryID, params, options) {
        const { id } = params;
        return this._client.put((0, path_1.path) `/api/ledger_account_categories/${id}/ledger_account_categories/${subCategoryID}`, { ...options, headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]) });
    }
    /**
     * Remove a ledger account from a ledger account category.
     *
     * @example
     * ```ts
     * await client.ledgerAccountCategories.removeLedgerAccount(
     *   'ledger_account_id',
     *   { id: 'id' },
     * );
     * ```
     */
    removeLedgerAccount(ledgerAccountID, params, options) {
        const { id } = params;
        return this._client.delete((0, path_1.path) `/api/ledger_account_categories/${id}/ledger_accounts/${ledgerAccountID}`, { ...options, headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]) });
    }
    /**
     * Delete a ledger account category from a ledger account category.
     *
     * @example
     * ```ts
     * await client.ledgerAccountCategories.removeNestedCategory(
     *   'sub_category_id',
     *   { id: 'id' },
     * );
     * ```
     */
    removeNestedCategory(subCategoryID, params, options) {
        const { id } = params;
        return this._client.delete((0, path_1.path) `/api/ledger_account_categories/${id}/ledger_account_categories/${subCategoryID}`, { ...options, headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]) });
    }
}
exports.LedgerAccountCategories = LedgerAccountCategories;
