// modern-treasury/resources/virtual-accounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { Page } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class VirtualAccounts extends APIResource {
    /**
     * create virtual_account
     */
    create(body, options) {
        return this._client.post('/api/virtual_accounts', { body, ...options });
    }
    /**
     * get virtual_account
     */
    retrieve(id, options) {
        return this._client.get(path `/api/virtual_accounts/${id}`, options);
    }
    /**
     * update virtual_account
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/virtual_accounts/${id}`, { body, ...options });
    }
    /**
     * Get a list of virtual accounts.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/virtual_accounts', (Page), { query, ...options });
    }
    /**
     * delete virtual_account
     */
    delete(id, options) {
        return this._client.delete(path `/api/virtual_accounts/${id}`, options);
    }
}
