// modern-treasury/resources/virtual-accounts.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualAccounts = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class VirtualAccounts extends resource_1.APIResource {
    /**
     * create virtual_account
     */
    create(body, options) {
        return this._client.post('/api/virtual_accounts', { body, ...options });
    }
    /**
     * get virtual_account
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/virtual_accounts/${id}`, options);
    }
    /**
     * update virtual_account
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/virtual_accounts/${id}`, { body, ...options });
    }
    /**
     * Get a list of virtual accounts.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/virtual_accounts', (pagination_1.Page), { query, ...options });
    }
    /**
     * delete virtual_account
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/virtual_accounts/${id}`, options);
    }
}
exports.VirtualAccounts = VirtualAccounts;
