// modern-treasury/resources/ledger-account-settlements/ledger-account-settlements.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerAccountSettlements = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const AccountEntriesAPI = tslib_1.__importStar(require("./account-entries.js"));
const account_entries_1 = require("./account-entries.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class LedgerAccountSettlements extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.accountEntries = new AccountEntriesAPI.AccountEntries(this._client);
    }
    /**
     * Create a ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.create({
     *     contra_ledger_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     settled_ledger_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_account_settlements', { body, ...options });
    }
    /**
     * Get details on a single ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/ledger_account_settlements/${id}`, options);
    }
    /**
     * Update the details of a ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/ledger_account_settlements/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger account settlements.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccountSettlement of client.ledgerAccountSettlements.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_account_settlements', (pagination_1.Page), {
            query,
            ...options,
        });
    }
}
exports.LedgerAccountSettlements = LedgerAccountSettlements;
LedgerAccountSettlements.AccountEntries = account_entries_1.AccountEntries;
