// modern-treasury/resources/ledger-account-settlements/ledger-account-settlements.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as AccountEntriesAPI from "./account-entries.mjs";
import { AccountEntries } from "./account-entries.mjs";
import { Page } from "../../core/pagination.mjs";
import { path } from "../../internal/utils/path.mjs";
export class LedgerAccountSettlements extends APIResource {
    constructor() {
        super(...arguments);
        this.accountEntries = new AccountEntriesAPI.AccountEntries(this._client);
    }
    /**
     * Create a ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.create({
     *     contra_ledger_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     settled_ledger_account_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_account_settlements', { body, ...options });
    }
    /**
     * Get details on a single ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/ledger_account_settlements/${id}`, options);
    }
    /**
     * Update the details of a ledger account settlement.
     *
     * @example
     * ```ts
     * const ledgerAccountSettlement =
     *   await client.ledgerAccountSettlements.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/ledger_account_settlements/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger account settlements.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerAccountSettlement of client.ledgerAccountSettlements.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_account_settlements', (Page), {
            query,
            ...options,
        });
    }
}
LedgerAccountSettlements.AccountEntries = AccountEntries;
