// modern-treasury/client.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var _ModernTreasury_instances, _a, _ModernTreasury_encoder, _ModernTreasury_baseURLOverridden;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "./internal/tslib.mjs";
import { uuid4 } from "./internal/utils/uuid.mjs";
import { validatePositiveInteger, isAbsoluteURL, safeJSON } from "./internal/utils/values.mjs";
import { sleep } from "./internal/utils/sleep.mjs";
import { castToError, isAbortError } from "./internal/errors.mjs";
import { getPlatformHeaders } from "./internal/detect-platform.mjs";
import * as Shims from "./internal/shims.mjs";
import * as Opts from "./internal/request-options.mjs";
import * as qs from "./internal/qs/index.mjs";
import { VERSION } from "./version.mjs";
import * as Errors from "./core/error.mjs";
import * as Pagination from "./core/pagination.mjs";
import * as Uploads from "./core/uploads.mjs";
import * as API from "./resources/index.mjs";
import { APIPromise } from "./core/api-promise.mjs";
import { AccountCollectionFlows, } from "./resources/account-collection-flows.mjs";
import { AccountDetails, } from "./resources/account-details.mjs";
import { BulkRequests, } from "./resources/bulk-requests.mjs";
import { BulkResults } from "./resources/bulk-results.mjs";
import { ConnectionLegalEntities, } from "./resources/connection-legal-entities.mjs";
import { Connections } from "./resources/connections.mjs";
import { Counterparties, } from "./resources/counterparties.mjs";
import { Documents, } from "./resources/documents.mjs";
import { Events } from "./resources/events.mjs";
import { ExpectedPayments, } from "./resources/expected-payments.mjs";
import { ExternalAccounts, } from "./resources/external-accounts.mjs";
import { ForeignExchangeQuotes, } from "./resources/foreign-exchange-quotes.mjs";
import { IncomingPaymentDetails, } from "./resources/incoming-payment-details.mjs";
import { LedgerAccountBalanceMonitors, } from "./resources/ledger-account-balance-monitors.mjs";
import { LedgerAccountCategories, } from "./resources/ledger-account-categories.mjs";
import { LedgerAccountStatements, } from "./resources/ledger-account-statements.mjs";
import { LedgerAccounts, } from "./resources/ledger-accounts.mjs";
import { LedgerEntries, } from "./resources/ledger-entries.mjs";
import { Ledgers, } from "./resources/ledgers.mjs";
import { LegalEntities, } from "./resources/legal-entities.mjs";
import { LegalEntityAssociations, } from "./resources/legal-entity-associations.mjs";
import { LineItems, } from "./resources/line-items.mjs";
import { PaymentActions, } from "./resources/payment-actions.mjs";
import { PaymentFlows, } from "./resources/payment-flows.mjs";
import { PaymentReferences, } from "./resources/payment-references.mjs";
import { Returns, } from "./resources/returns.mjs";
import { RoutingDetails, } from "./resources/routing-details.mjs";
import { Validations, } from "./resources/validations.mjs";
import { VirtualAccounts, } from "./resources/virtual-accounts.mjs";
import { InternalAccounts, } from "./resources/internal-accounts/internal-accounts.mjs";
import { Invoices, } from "./resources/invoices/invoices.mjs";
import { LedgerAccountSettlements, } from "./resources/ledger-account-settlements/ledger-account-settlements.mjs";
import { LedgerTransactions, } from "./resources/ledger-transactions/ledger-transactions.mjs";
import { PaymentOrders, } from "./resources/payment-orders/payment-orders.mjs";
import { Transactions, } from "./resources/transactions/transactions.mjs";
import { buildHeaders } from "./internal/headers.mjs";
import { toBase64 } from "./internal/utils/base64.mjs";
import { readEnv } from "./internal/utils/env.mjs";
import { formatRequestDetails, loggerFor, parseLogLevel, } from "./internal/utils/log.mjs";
import { isEmptyObj } from "./internal/utils/values.mjs";
/**
 * API Client for interfacing with the Modern Treasury API.
 */
export class ModernTreasury {
    /**
     * API Client for interfacing with the Modern Treasury API.
     *
     * @param {string | undefined} [opts.apiKey=process.env['MODERN_TREASURY_API_KEY'] ?? undefined]
     * @param {string | undefined} [opts.organizationID=process.env['MODERN_TREASURY_ORGANIZATION_ID'] ?? undefined]
     * @param {string | null | undefined} [opts.webhookKey=process.env['MODERN_TREASURY_WEBHOOK_KEY'] ?? null]
     * @param {string} [opts.baseURL=process.env['MODERN_TREASURY_BASE_URL'] ?? https://app.moderntreasury.com] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {MergedRequestInit} [opts.fetchOptions] - Additional `RequestInit` options to be passed to `fetch` calls.
     * @param {Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {HeadersLike} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Record<string, string | undefined>} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = readEnv('MODERN_TREASURY_BASE_URL'), apiKey = readEnv('MODERN_TREASURY_API_KEY'), organizationID = readEnv('MODERN_TREASURY_ORGANIZATION_ID'), webhookKey = readEnv('MODERN_TREASURY_WEBHOOK_KEY') ?? null, ...opts } = {}) {
        _ModernTreasury_instances.add(this);
        _ModernTreasury_encoder.set(this, void 0);
        this.webhooks = new API.Webhooks(this);
        this.connections = new API.Connections(this);
        this.counterparties = new API.Counterparties(this);
        this.events = new API.Events(this);
        this.expectedPayments = new API.ExpectedPayments(this);
        this.externalAccounts = new API.ExternalAccounts(this);
        this.incomingPaymentDetails = new API.IncomingPaymentDetails(this);
        this.invoices = new API.Invoices(this);
        this.documents = new API.Documents(this);
        this.accountCollectionFlows = new API.AccountCollectionFlows(this);
        this.accountDetails = new API.AccountDetails(this);
        this.routingDetails = new API.RoutingDetails(this);
        this.internalAccounts = new API.InternalAccounts(this);
        this.ledgers = new API.Ledgers(this);
        this.ledgerAccountCategories = new API.LedgerAccountCategories(this);
        this.ledgerAccounts = new API.LedgerAccounts(this);
        this.ledgerAccountBalanceMonitors = new API.LedgerAccountBalanceMonitors(this);
        this.ledgerAccountStatements = new API.LedgerAccountStatements(this);
        this.ledgerEntries = new API.LedgerEntries(this);
        this.ledgerTransactions = new API.LedgerTransactions(this);
        this.lineItems = new API.LineItems(this);
        this.paymentFlows = new API.PaymentFlows(this);
        this.paymentOrders = new API.PaymentOrders(this);
        this.paymentReferences = new API.PaymentReferences(this);
        this.returns = new API.Returns(this);
        this.transactions = new API.Transactions(this);
        this.validations = new API.Validations(this);
        this.virtualAccounts = new API.VirtualAccounts(this);
        this.bulkRequests = new API.BulkRequests(this);
        this.bulkResults = new API.BulkResults(this);
        this.ledgerAccountSettlements = new API.LedgerAccountSettlements(this);
        this.foreignExchangeQuotes = new API.ForeignExchangeQuotes(this);
        this.connectionLegalEntities = new API.ConnectionLegalEntities(this);
        this.legalEntities = new API.LegalEntities(this);
        this.legalEntityAssociations = new API.LegalEntityAssociations(this);
        this.paymentActions = new API.PaymentActions(this);
        if (apiKey === undefined) {
            throw new Errors.ModernTreasuryError("The MODERN_TREASURY_API_KEY environment variable is missing or empty; either provide it, or instantiate the ModernTreasury client with an apiKey option, like new ModernTreasury({ apiKey: 'My API Key' }).");
        }
        if (organizationID === undefined) {
            throw new Errors.ModernTreasuryError("The MODERN_TREASURY_ORGANIZATION_ID environment variable is missing or empty; either provide it, or instantiate the ModernTreasury client with an organizationID option, like new ModernTreasury({ organizationID: 'my-organization-ID' }).");
        }
        const options = {
            apiKey,
            organizationID,
            webhookKey,
            ...opts,
            baseURL: baseURL || `https://app.moderntreasury.com`,
        };
        this.baseURL = options.baseURL;
        this.timeout = options.timeout ?? _a.DEFAULT_TIMEOUT /* 1 minute */;
        this.logger = options.logger ?? console;
        const defaultLogLevel = 'warn';
        // Set default logLevel early so that we can log a warning in parseLogLevel.
        this.logLevel = defaultLogLevel;
        this.logLevel =
            parseLogLevel(options.logLevel, 'ClientOptions.logLevel', this) ??
                parseLogLevel(readEnv('MODERN_TREASURY_LOG'), "process.env['MODERN_TREASURY_LOG']", this) ??
                defaultLogLevel;
        this.fetchOptions = options.fetchOptions;
        this.maxRetries = options.maxRetries ?? 2;
        this.fetch = options.fetch ?? Shims.getDefaultFetch();
        __classPrivateFieldSet(this, _ModernTreasury_encoder, Opts.FallbackEncoder, "f");
        this._options = options;
        this.idempotencyHeader = 'Idempotency-Key';
        this.apiKey = apiKey;
        this.organizationID = organizationID;
        this.webhookKey = webhookKey;
    }
    /**
     * Create a new client instance re-using the same options given to the current client with optional overriding.
     */
    withOptions(options) {
        const client = new this.constructor({
            ...this._options,
            baseURL: this.baseURL,
            maxRetries: this.maxRetries,
            timeout: this.timeout,
            logger: this.logger,
            logLevel: this.logLevel,
            fetch: this.fetch,
            fetchOptions: this.fetchOptions,
            apiKey: this.apiKey,
            organizationID: this.organizationID,
            webhookKey: this.webhookKey,
            ...options,
        });
        return client;
    }
    /**
     * A test endpoint often used to confirm credentials and headers are being passed
     * in correctly.
     */
    ping(options) {
        return this.get('/api/ping', options);
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    validateHeaders({ values, nulls }) {
        return;
    }
    async authHeaders(opts) {
        if (!this.organizationID) {
            return undefined;
        }
        if (!this.apiKey) {
            return undefined;
        }
        const credentials = `${this.organizationID}:${this.apiKey}`;
        const Authorization = `Basic ${toBase64(credentials)}`;
        return buildHeaders([{ Authorization }]);
    }
    stringifyQuery(query) {
        return qs.stringify(query, { arrayFormat: 'brackets' });
    }
    getUserAgent() {
        return `${this.constructor.name}/JS ${VERSION}`;
    }
    defaultIdempotencyKey() {
        return `stainless-node-retry-${uuid4()}`;
    }
    makeStatusError(status, error, message, headers) {
        return Errors.APIError.generate(status, error, message, headers);
    }
    buildURL(path, query, defaultBaseURL) {
        const baseURL = (!__classPrivateFieldGet(this, _ModernTreasury_instances, "m", _ModernTreasury_baseURLOverridden).call(this) && defaultBaseURL) || this.baseURL;
        const url = isAbsoluteURL(path) ?
            new URL(path)
            : new URL(baseURL + (baseURL.endsWith('/') && path.startsWith('/') ? path.slice(1) : path));
        const defaultQuery = this.defaultQuery();
        if (!isEmptyObj(defaultQuery)) {
            query = { ...defaultQuery, ...query };
        }
        if (typeof query === 'object' && query && !Array.isArray(query)) {
            url.search = this.stringifyQuery(query);
        }
        return url.toString();
    }
    /**
     * Used as a callback for mutating the given `FinalRequestOptions` object.
     */
    async prepareOptions(options) { }
    /**
     * Used as a callback for mutating the given `RequestInit` object.
     *
     * This is useful for cases where you want to add certain headers based off of
     * the request properties, e.g. `method` or `url`.
     */
    async prepareRequest(request, { url, options }) { }
    get(path, opts) {
        return this.methodRequest('get', path, opts);
    }
    post(path, opts) {
        return this.methodRequest('post', path, opts);
    }
    patch(path, opts) {
        return this.methodRequest('patch', path, opts);
    }
    put(path, opts) {
        return this.methodRequest('put', path, opts);
    }
    delete(path, opts) {
        return this.methodRequest('delete', path, opts);
    }
    methodRequest(method, path, opts) {
        return this.request(Promise.resolve(opts).then((opts) => {
            return { method, path, ...opts };
        }));
    }
    request(options, remainingRetries = null) {
        return new APIPromise(this, this.makeRequest(options, remainingRetries, undefined));
    }
    async makeRequest(optionsInput, retriesRemaining, retryOfRequestLogID) {
        const options = await optionsInput;
        const maxRetries = options.maxRetries ?? this.maxRetries;
        if (retriesRemaining == null) {
            retriesRemaining = maxRetries;
        }
        await this.prepareOptions(options);
        const { req, url, timeout } = await this.buildRequest(options, {
            retryCount: maxRetries - retriesRemaining,
        });
        await this.prepareRequest(req, { url, options });
        /** Not an API request ID, just for correlating local log entries. */
        const requestLogID = 'log_' + ((Math.random() * (1 << 24)) | 0).toString(16).padStart(6, '0');
        const retryLogStr = retryOfRequestLogID === undefined ? '' : `, retryOf: ${retryOfRequestLogID}`;
        const startTime = Date.now();
        loggerFor(this).debug(`[${requestLogID}] sending request`, formatRequestDetails({
            retryOfRequestLogID,
            method: options.method,
            url,
            options,
            headers: req.headers,
        }));
        if (options.signal?.aborted) {
            throw new Errors.APIUserAbortError();
        }
        const controller = new AbortController();
        const response = await this.fetchWithTimeout(url, req, timeout, controller).catch(castToError);
        const headersTime = Date.now();
        if (response instanceof globalThis.Error) {
            const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;
            if (options.signal?.aborted) {
                throw new Errors.APIUserAbortError();
            }
            // detect native connection timeout errors
            // deno throws "TypeError: error sending request for url (https://example/): client error (Connect): tcp connect error: Operation timed out (os error 60): Operation timed out (os error 60)"
            // undici throws "TypeError: fetch failed" with cause "ConnectTimeoutError: Connect Timeout Error (attempted address: example:443, timeout: 1ms)"
            // others do not provide enough information to distinguish timeouts from other connection errors
            const isTimeout = isAbortError(response) ||
                /timed? ?out/i.test(String(response) + ('cause' in response ? String(response.cause) : ''));
            if (retriesRemaining) {
                loggerFor(this).info(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - ${retryMessage}`);
                loggerFor(this).debug(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (${retryMessage})`, formatRequestDetails({
                    retryOfRequestLogID,
                    url,
                    durationMs: headersTime - startTime,
                    message: response.message,
                }));
                return this.retryRequest(options, retriesRemaining, retryOfRequestLogID ?? requestLogID);
            }
            loggerFor(this).info(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - error; no more retries left`);
            loggerFor(this).debug(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (error; no more retries left)`, formatRequestDetails({
                retryOfRequestLogID,
                url,
                durationMs: headersTime - startTime,
                message: response.message,
            }));
            if (isTimeout) {
                throw new Errors.APIConnectionTimeoutError();
            }
            throw new Errors.APIConnectionError({ cause: response });
        }
        const responseInfo = `[${requestLogID}${retryLogStr}] ${req.method} ${url} ${response.ok ? 'succeeded' : 'failed'} with status ${response.status} in ${headersTime - startTime}ms`;
        if (!response.ok) {
            const shouldRetry = await this.shouldRetry(response);
            if (retriesRemaining && shouldRetry) {
                const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;
                // We don't need the body of this response.
                await Shims.CancelReadableStream(response.body);
                loggerFor(this).info(`${responseInfo} - ${retryMessage}`);
                loggerFor(this).debug(`[${requestLogID}] response error (${retryMessage})`, formatRequestDetails({
                    retryOfRequestLogID,
                    url: response.url,
                    status: response.status,
                    headers: response.headers,
                    durationMs: headersTime - startTime,
                }));
                return this.retryRequest(options, retriesRemaining, retryOfRequestLogID ?? requestLogID, response.headers);
            }
            const retryMessage = shouldRetry ? `error; no more retries left` : `error; not retryable`;
            loggerFor(this).info(`${responseInfo} - ${retryMessage}`);
            const errText = await response.text().catch((err) => castToError(err).message);
            const errJSON = safeJSON(errText);
            const errMessage = errJSON ? undefined : errText;
            loggerFor(this).debug(`[${requestLogID}] response error (${retryMessage})`, formatRequestDetails({
                retryOfRequestLogID,
                url: response.url,
                status: response.status,
                headers: response.headers,
                message: errMessage,
                durationMs: Date.now() - startTime,
            }));
            const err = this.makeStatusError(response.status, errJSON, errMessage, response.headers);
            throw err;
        }
        loggerFor(this).info(responseInfo);
        loggerFor(this).debug(`[${requestLogID}] response start`, formatRequestDetails({
            retryOfRequestLogID,
            url: response.url,
            status: response.status,
            headers: response.headers,
            durationMs: headersTime - startTime,
        }));
        return { response, options, controller, requestLogID, retryOfRequestLogID, startTime };
    }
    getAPIList(path, Page, opts) {
        return this.requestAPIList(Page, { method: 'get', path, ...opts });
    }
    requestAPIList(Page, options) {
        const request = this.makeRequest(options, null, undefined);
        return new Pagination.PagePromise(this, request, Page);
    }
    async fetchWithTimeout(url, init, ms, controller) {
        const { signal, method, ...options } = init || {};
        if (signal)
            signal.addEventListener('abort', () => controller.abort());
        const timeout = setTimeout(() => controller.abort(), ms);
        const isReadableBody = (globalThis.ReadableStream && options.body instanceof globalThis.ReadableStream) ||
            (typeof options.body === 'object' && options.body !== null && Symbol.asyncIterator in options.body);
        const fetchOptions = {
            signal: controller.signal,
            ...(isReadableBody ? { duplex: 'half' } : {}),
            method: 'GET',
            ...options,
        };
        if (method) {
            // Custom methods like 'patch' need to be uppercased
            // See https://github.com/nodejs/undici/issues/2294
            fetchOptions.method = method.toUpperCase();
        }
        try {
            // use undefined this binding; fetch errors if bound to something else in browser/cloudflare
            return await this.fetch.call(undefined, url, fetchOptions);
        }
        finally {
            clearTimeout(timeout);
        }
    }
    async shouldRetry(response) {
        // Note this is not a standard header.
        const shouldRetryHeader = response.headers.get('x-should-retry');
        // If the server explicitly says whether or not to retry, obey.
        if (shouldRetryHeader === 'true')
            return true;
        if (shouldRetryHeader === 'false')
            return false;
        // Retry on request timeouts.
        if (response.status === 408)
            return true;
        // Retry on lock timeouts.
        if (response.status === 409)
            return true;
        // Retry on rate limits.
        if (response.status === 429)
            return true;
        // Retry internal errors.
        if (response.status >= 500)
            return true;
        return false;
    }
    async retryRequest(options, retriesRemaining, requestLogID, responseHeaders) {
        let timeoutMillis;
        // Note the `retry-after-ms` header may not be standard, but is a good idea and we'd like proactive support for it.
        const retryAfterMillisHeader = responseHeaders?.get('retry-after-ms');
        if (retryAfterMillisHeader) {
            const timeoutMs = parseFloat(retryAfterMillisHeader);
            if (!Number.isNaN(timeoutMs)) {
                timeoutMillis = timeoutMs;
            }
        }
        // About the Retry-After header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
        const retryAfterHeader = responseHeaders?.get('retry-after');
        if (retryAfterHeader && !timeoutMillis) {
            const timeoutSeconds = parseFloat(retryAfterHeader);
            if (!Number.isNaN(timeoutSeconds)) {
                timeoutMillis = timeoutSeconds * 1000;
            }
            else {
                timeoutMillis = Date.parse(retryAfterHeader) - Date.now();
            }
        }
        // If the API asks us to wait a certain amount of time (and it's a reasonable amount),
        // just do what it says, but otherwise calculate a default
        if (!(timeoutMillis && 0 <= timeoutMillis && timeoutMillis < 60 * 1000)) {
            const maxRetries = options.maxRetries ?? this.maxRetries;
            timeoutMillis = this.calculateDefaultRetryTimeoutMillis(retriesRemaining, maxRetries);
        }
        await sleep(timeoutMillis);
        return this.makeRequest(options, retriesRemaining - 1, requestLogID);
    }
    calculateDefaultRetryTimeoutMillis(retriesRemaining, maxRetries) {
        const initialRetryDelay = 0.5;
        const maxRetryDelay = 8.0;
        const numRetries = maxRetries - retriesRemaining;
        // Apply exponential backoff, but not more than the max.
        const sleepSeconds = Math.min(initialRetryDelay * Math.pow(2, numRetries), maxRetryDelay);
        // Apply some jitter, take up to at most 25 percent of the retry time.
        const jitter = 1 - Math.random() * 0.25;
        return sleepSeconds * jitter * 1000;
    }
    async buildRequest(inputOptions, { retryCount = 0 } = {}) {
        const options = { ...inputOptions };
        const { method, path, query, defaultBaseURL } = options;
        const url = this.buildURL(path, query, defaultBaseURL);
        if ('timeout' in options)
            validatePositiveInteger('timeout', options.timeout);
        options.timeout = options.timeout ?? this.timeout;
        const { bodyHeaders, body } = this.buildBody({ options });
        const reqHeaders = await this.buildHeaders({ options: inputOptions, method, bodyHeaders, retryCount });
        const req = {
            method,
            headers: reqHeaders,
            ...(options.signal && { signal: options.signal }),
            ...(globalThis.ReadableStream &&
                body instanceof globalThis.ReadableStream && { duplex: 'half' }),
            ...(body && { body }),
            ...(this.fetchOptions ?? {}),
            ...(options.fetchOptions ?? {}),
        };
        return { req, url, timeout: options.timeout };
    }
    async buildHeaders({ options, method, bodyHeaders, retryCount, }) {
        let idempotencyHeaders = {};
        if (this.idempotencyHeader && method !== 'get') {
            if (!options.idempotencyKey)
                options.idempotencyKey = this.defaultIdempotencyKey();
            idempotencyHeaders[this.idempotencyHeader] = options.idempotencyKey;
        }
        const headers = buildHeaders([
            idempotencyHeaders,
            {
                Accept: 'application/json',
                'User-Agent': this.getUserAgent(),
                'X-Stainless-Retry-Count': String(retryCount),
                ...(options.timeout ? { 'X-Stainless-Timeout': String(Math.trunc(options.timeout / 1000)) } : {}),
                ...getPlatformHeaders(),
            },
            await this.authHeaders(options),
            this._options.defaultHeaders,
            bodyHeaders,
            options.headers,
        ]);
        this.validateHeaders(headers);
        return headers.values;
    }
    buildBody({ options: { body, headers: rawHeaders } }) {
        if (!body) {
            return { bodyHeaders: undefined, body: undefined };
        }
        const headers = buildHeaders([rawHeaders]);
        if (
        // Pass raw type verbatim
        ArrayBuffer.isView(body) ||
            body instanceof ArrayBuffer ||
            body instanceof DataView ||
            (typeof body === 'string' &&
                // Preserve legacy string encoding behavior for now
                headers.values.has('content-type')) ||
            // `Blob` is superset of `File`
            (globalThis.Blob && body instanceof globalThis.Blob) ||
            // `FormData` -> `multipart/form-data`
            body instanceof FormData ||
            // `URLSearchParams` -> `application/x-www-form-urlencoded`
            body instanceof URLSearchParams ||
            // Send chunked stream (each chunk has own `length`)
            (globalThis.ReadableStream && body instanceof globalThis.ReadableStream)) {
            return { bodyHeaders: undefined, body: body };
        }
        else if (typeof body === 'object' &&
            (Symbol.asyncIterator in body ||
                (Symbol.iterator in body && 'next' in body && typeof body.next === 'function'))) {
            return { bodyHeaders: undefined, body: Shims.ReadableStreamFrom(body) };
        }
        else {
            return __classPrivateFieldGet(this, _ModernTreasury_encoder, "f").call(this, { body, headers });
        }
    }
}
_a = ModernTreasury, _ModernTreasury_encoder = new WeakMap(), _ModernTreasury_instances = new WeakSet(), _ModernTreasury_baseURLOverridden = function _ModernTreasury_baseURLOverridden() {
    return this.baseURL !== 'https://app.moderntreasury.com';
};
ModernTreasury.ModernTreasury = _a;
ModernTreasury.DEFAULT_TIMEOUT = 60000; // 1 minute
ModernTreasury.ModernTreasuryError = Errors.ModernTreasuryError;
ModernTreasury.APIError = Errors.APIError;
ModernTreasury.APIConnectionError = Errors.APIConnectionError;
ModernTreasury.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
ModernTreasury.APIUserAbortError = Errors.APIUserAbortError;
ModernTreasury.NotFoundError = Errors.NotFoundError;
ModernTreasury.ConflictError = Errors.ConflictError;
ModernTreasury.RateLimitError = Errors.RateLimitError;
ModernTreasury.BadRequestError = Errors.BadRequestError;
ModernTreasury.AuthenticationError = Errors.AuthenticationError;
ModernTreasury.InternalServerError = Errors.InternalServerError;
ModernTreasury.PermissionDeniedError = Errors.PermissionDeniedError;
ModernTreasury.UnprocessableEntityError = Errors.UnprocessableEntityError;
ModernTreasury.toFile = Uploads.toFile;
ModernTreasury.Connections = Connections;
ModernTreasury.Counterparties = Counterparties;
ModernTreasury.Events = Events;
ModernTreasury.ExpectedPayments = ExpectedPayments;
ModernTreasury.ExternalAccounts = ExternalAccounts;
ModernTreasury.IncomingPaymentDetails = IncomingPaymentDetails;
ModernTreasury.Invoices = Invoices;
ModernTreasury.Documents = Documents;
ModernTreasury.AccountCollectionFlows = AccountCollectionFlows;
ModernTreasury.AccountDetails = AccountDetails;
ModernTreasury.RoutingDetails = RoutingDetails;
ModernTreasury.InternalAccounts = InternalAccounts;
ModernTreasury.Ledgers = Ledgers;
ModernTreasury.LedgerAccountCategories = LedgerAccountCategories;
ModernTreasury.LedgerAccounts = LedgerAccounts;
ModernTreasury.LedgerAccountBalanceMonitors = LedgerAccountBalanceMonitors;
ModernTreasury.LedgerAccountStatements = LedgerAccountStatements;
ModernTreasury.LedgerEntries = LedgerEntries;
ModernTreasury.LedgerTransactions = LedgerTransactions;
ModernTreasury.LineItems = LineItems;
ModernTreasury.PaymentFlows = PaymentFlows;
ModernTreasury.PaymentOrders = PaymentOrders;
ModernTreasury.PaymentReferences = PaymentReferences;
ModernTreasury.Returns = Returns;
ModernTreasury.Transactions = Transactions;
ModernTreasury.Validations = Validations;
ModernTreasury.VirtualAccounts = VirtualAccounts;
ModernTreasury.BulkRequests = BulkRequests;
ModernTreasury.BulkResults = BulkResults;
ModernTreasury.LedgerAccountSettlements = LedgerAccountSettlements;
ModernTreasury.ForeignExchangeQuotes = ForeignExchangeQuotes;
ModernTreasury.ConnectionLegalEntities = ConnectionLegalEntities;
ModernTreasury.LegalEntities = LegalEntities;
ModernTreasury.LegalEntityAssociations = LegalEntityAssociations;
ModernTreasury.PaymentActions = PaymentActions;
