// modern-treasury/client.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var _ModernTreasury_instances, _a, _ModernTreasury_encoder, _ModernTreasury_baseURLOverridden;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModernTreasury = void 0;
const tslib_1 = require("./internal/tslib.js");
const uuid_1 = require("./internal/utils/uuid.js");
const values_1 = require("./internal/utils/values.js");
const sleep_1 = require("./internal/utils/sleep.js");
const errors_1 = require("./internal/errors.js");
const detect_platform_1 = require("./internal/detect-platform.js");
const Shims = tslib_1.__importStar(require("./internal/shims.js"));
const Opts = tslib_1.__importStar(require("./internal/request-options.js"));
const qs = tslib_1.__importStar(require("./internal/qs/index.js"));
const version_1 = require("./version.js");
const Errors = tslib_1.__importStar(require("./core/error.js"));
const Pagination = tslib_1.__importStar(require("./core/pagination.js"));
const Uploads = tslib_1.__importStar(require("./core/uploads.js"));
const API = tslib_1.__importStar(require("./resources/index.js"));
const api_promise_1 = require("./core/api-promise.js");
const account_collection_flows_1 = require("./resources/account-collection-flows.js");
const account_details_1 = require("./resources/account-details.js");
const bulk_requests_1 = require("./resources/bulk-requests.js");
const bulk_results_1 = require("./resources/bulk-results.js");
const connection_legal_entities_1 = require("./resources/connection-legal-entities.js");
const connections_1 = require("./resources/connections.js");
const counterparties_1 = require("./resources/counterparties.js");
const documents_1 = require("./resources/documents.js");
const events_1 = require("./resources/events.js");
const expected_payments_1 = require("./resources/expected-payments.js");
const external_accounts_1 = require("./resources/external-accounts.js");
const foreign_exchange_quotes_1 = require("./resources/foreign-exchange-quotes.js");
const incoming_payment_details_1 = require("./resources/incoming-payment-details.js");
const ledger_account_balance_monitors_1 = require("./resources/ledger-account-balance-monitors.js");
const ledger_account_categories_1 = require("./resources/ledger-account-categories.js");
const ledger_account_statements_1 = require("./resources/ledger-account-statements.js");
const ledger_accounts_1 = require("./resources/ledger-accounts.js");
const ledger_entries_1 = require("./resources/ledger-entries.js");
const ledgers_1 = require("./resources/ledgers.js");
const legal_entities_1 = require("./resources/legal-entities.js");
const legal_entity_associations_1 = require("./resources/legal-entity-associations.js");
const line_items_1 = require("./resources/line-items.js");
const payment_actions_1 = require("./resources/payment-actions.js");
const payment_flows_1 = require("./resources/payment-flows.js");
const payment_references_1 = require("./resources/payment-references.js");
const returns_1 = require("./resources/returns.js");
const routing_details_1 = require("./resources/routing-details.js");
const validations_1 = require("./resources/validations.js");
const virtual_accounts_1 = require("./resources/virtual-accounts.js");
const internal_accounts_1 = require("./resources/internal-accounts/internal-accounts.js");
const invoices_1 = require("./resources/invoices/invoices.js");
const ledger_account_settlements_1 = require("./resources/ledger-account-settlements/ledger-account-settlements.js");
const ledger_transactions_1 = require("./resources/ledger-transactions/ledger-transactions.js");
const payment_orders_1 = require("./resources/payment-orders/payment-orders.js");
const transactions_1 = require("./resources/transactions/transactions.js");
const headers_1 = require("./internal/headers.js");
const base64_1 = require("./internal/utils/base64.js");
const env_1 = require("./internal/utils/env.js");
const log_1 = require("./internal/utils/log.js");
const values_2 = require("./internal/utils/values.js");
/**
 * API Client for interfacing with the Modern Treasury API.
 */
class ModernTreasury {
    /**
     * API Client for interfacing with the Modern Treasury API.
     *
     * @param {string | undefined} [opts.apiKey=process.env['MODERN_TREASURY_API_KEY'] ?? undefined]
     * @param {string | undefined} [opts.organizationID=process.env['MODERN_TREASURY_ORGANIZATION_ID'] ?? undefined]
     * @param {string | null | undefined} [opts.webhookKey=process.env['MODERN_TREASURY_WEBHOOK_KEY'] ?? null]
     * @param {string} [opts.baseURL=process.env['MODERN_TREASURY_BASE_URL'] ?? https://app.moderntreasury.com] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {MergedRequestInit} [opts.fetchOptions] - Additional `RequestInit` options to be passed to `fetch` calls.
     * @param {Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {HeadersLike} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Record<string, string | undefined>} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = (0, env_1.readEnv)('MODERN_TREASURY_BASE_URL'), apiKey = (0, env_1.readEnv)('MODERN_TREASURY_API_KEY'), organizationID = (0, env_1.readEnv)('MODERN_TREASURY_ORGANIZATION_ID'), webhookKey = (0, env_1.readEnv)('MODERN_TREASURY_WEBHOOK_KEY') ?? null, ...opts } = {}) {
        _ModernTreasury_instances.add(this);
        _ModernTreasury_encoder.set(this, void 0);
        this.webhooks = new API.Webhooks(this);
        this.connections = new API.Connections(this);
        this.counterparties = new API.Counterparties(this);
        this.events = new API.Events(this);
        this.expectedPayments = new API.ExpectedPayments(this);
        this.externalAccounts = new API.ExternalAccounts(this);
        this.incomingPaymentDetails = new API.IncomingPaymentDetails(this);
        this.invoices = new API.Invoices(this);
        this.documents = new API.Documents(this);
        this.accountCollectionFlows = new API.AccountCollectionFlows(this);
        this.accountDetails = new API.AccountDetails(this);
        this.routingDetails = new API.RoutingDetails(this);
        this.internalAccounts = new API.InternalAccounts(this);
        this.ledgers = new API.Ledgers(this);
        this.ledgerAccountCategories = new API.LedgerAccountCategories(this);
        this.ledgerAccounts = new API.LedgerAccounts(this);
        this.ledgerAccountBalanceMonitors = new API.LedgerAccountBalanceMonitors(this);
        this.ledgerAccountStatements = new API.LedgerAccountStatements(this);
        this.ledgerEntries = new API.LedgerEntries(this);
        this.ledgerTransactions = new API.LedgerTransactions(this);
        this.lineItems = new API.LineItems(this);
        this.paymentFlows = new API.PaymentFlows(this);
        this.paymentOrders = new API.PaymentOrders(this);
        this.paymentReferences = new API.PaymentReferences(this);
        this.returns = new API.Returns(this);
        this.transactions = new API.Transactions(this);
        this.validations = new API.Validations(this);
        this.virtualAccounts = new API.VirtualAccounts(this);
        this.bulkRequests = new API.BulkRequests(this);
        this.bulkResults = new API.BulkResults(this);
        this.ledgerAccountSettlements = new API.LedgerAccountSettlements(this);
        this.foreignExchangeQuotes = new API.ForeignExchangeQuotes(this);
        this.connectionLegalEntities = new API.ConnectionLegalEntities(this);
        this.legalEntities = new API.LegalEntities(this);
        this.legalEntityAssociations = new API.LegalEntityAssociations(this);
        this.paymentActions = new API.PaymentActions(this);
        if (apiKey === undefined) {
            throw new Errors.ModernTreasuryError("The MODERN_TREASURY_API_KEY environment variable is missing or empty; either provide it, or instantiate the ModernTreasury client with an apiKey option, like new ModernTreasury({ apiKey: 'My API Key' }).");
        }
        if (organizationID === undefined) {
            throw new Errors.ModernTreasuryError("The MODERN_TREASURY_ORGANIZATION_ID environment variable is missing or empty; either provide it, or instantiate the ModernTreasury client with an organizationID option, like new ModernTreasury({ organizationID: 'my-organization-ID' }).");
        }
        const options = {
            apiKey,
            organizationID,
            webhookKey,
            ...opts,
            baseURL: baseURL || `https://app.moderntreasury.com`,
        };
        this.baseURL = options.baseURL;
        this.timeout = options.timeout ?? _a.DEFAULT_TIMEOUT /* 1 minute */;
        this.logger = options.logger ?? console;
        const defaultLogLevel = 'warn';
        // Set default logLevel early so that we can log a warning in parseLogLevel.
        this.logLevel = defaultLogLevel;
        this.logLevel =
            (0, log_1.parseLogLevel)(options.logLevel, 'ClientOptions.logLevel', this) ??
                (0, log_1.parseLogLevel)((0, env_1.readEnv)('MODERN_TREASURY_LOG'), "process.env['MODERN_TREASURY_LOG']", this) ??
                defaultLogLevel;
        this.fetchOptions = options.fetchOptions;
        this.maxRetries = options.maxRetries ?? 2;
        this.fetch = options.fetch ?? Shims.getDefaultFetch();
        tslib_1.__classPrivateFieldSet(this, _ModernTreasury_encoder, Opts.FallbackEncoder, "f");
        this._options = options;
        this.idempotencyHeader = 'Idempotency-Key';
        this.apiKey = apiKey;
        this.organizationID = organizationID;
        this.webhookKey = webhookKey;
    }
    /**
     * Create a new client instance re-using the same options given to the current client with optional overriding.
     */
    withOptions(options) {
        const client = new this.constructor({
            ...this._options,
            baseURL: this.baseURL,
            maxRetries: this.maxRetries,
            timeout: this.timeout,
            logger: this.logger,
            logLevel: this.logLevel,
            fetch: this.fetch,
            fetchOptions: this.fetchOptions,
            apiKey: this.apiKey,
            organizationID: this.organizationID,
            webhookKey: this.webhookKey,
            ...options,
        });
        return client;
    }
    /**
     * A test endpoint often used to confirm credentials and headers are being passed
     * in correctly.
     */
    ping(options) {
        return this.get('/api/ping', options);
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    validateHeaders({ values, nulls }) {
        return;
    }
    async authHeaders(opts) {
        if (!this.organizationID) {
            return undefined;
        }
        if (!this.apiKey) {
            return undefined;
        }
        const credentials = `${this.organizationID}:${this.apiKey}`;
        const Authorization = `Basic ${(0, base64_1.toBase64)(credentials)}`;
        return (0, headers_1.buildHeaders)([{ Authorization }]);
    }
    stringifyQuery(query) {
        return qs.stringify(query, { arrayFormat: 'brackets' });
    }
    getUserAgent() {
        return `${this.constructor.name}/JS ${version_1.VERSION}`;
    }
    defaultIdempotencyKey() {
        return `stainless-node-retry-${(0, uuid_1.uuid4)()}`;
    }
    makeStatusError(status, error, message, headers) {
        return Errors.APIError.generate(status, error, message, headers);
    }
    buildURL(path, query, defaultBaseURL) {
        const baseURL = (!tslib_1.__classPrivateFieldGet(this, _ModernTreasury_instances, "m", _ModernTreasury_baseURLOverridden).call(this) && defaultBaseURL) || this.baseURL;
        const url = (0, values_1.isAbsoluteURL)(path) ?
            new URL(path)
            : new URL(baseURL + (baseURL.endsWith('/') && path.startsWith('/') ? path.slice(1) : path));
        const defaultQuery = this.defaultQuery();
        if (!(0, values_2.isEmptyObj)(defaultQuery)) {
            query = { ...defaultQuery, ...query };
        }
        if (typeof query === 'object' && query && !Array.isArray(query)) {
            url.search = this.stringifyQuery(query);
        }
        return url.toString();
    }
    /**
     * Used as a callback for mutating the given `FinalRequestOptions` object.
     */
    async prepareOptions(options) { }
    /**
     * Used as a callback for mutating the given `RequestInit` object.
     *
     * This is useful for cases where you want to add certain headers based off of
     * the request properties, e.g. `method` or `url`.
     */
    async prepareRequest(request, { url, options }) { }
    get(path, opts) {
        return this.methodRequest('get', path, opts);
    }
    post(path, opts) {
        return this.methodRequest('post', path, opts);
    }
    patch(path, opts) {
        return this.methodRequest('patch', path, opts);
    }
    put(path, opts) {
        return this.methodRequest('put', path, opts);
    }
    delete(path, opts) {
        return this.methodRequest('delete', path, opts);
    }
    methodRequest(method, path, opts) {
        return this.request(Promise.resolve(opts).then((opts) => {
            return { method, path, ...opts };
        }));
    }
    request(options, remainingRetries = null) {
        return new api_promise_1.APIPromise(this, this.makeRequest(options, remainingRetries, undefined));
    }
    async makeRequest(optionsInput, retriesRemaining, retryOfRequestLogID) {
        const options = await optionsInput;
        const maxRetries = options.maxRetries ?? this.maxRetries;
        if (retriesRemaining == null) {
            retriesRemaining = maxRetries;
        }
        await this.prepareOptions(options);
        const { req, url, timeout } = await this.buildRequest(options, {
            retryCount: maxRetries - retriesRemaining,
        });
        await this.prepareRequest(req, { url, options });
        /** Not an API request ID, just for correlating local log entries. */
        const requestLogID = 'log_' + ((Math.random() * (1 << 24)) | 0).toString(16).padStart(6, '0');
        const retryLogStr = retryOfRequestLogID === undefined ? '' : `, retryOf: ${retryOfRequestLogID}`;
        const startTime = Date.now();
        (0, log_1.loggerFor)(this).debug(`[${requestLogID}] sending request`, (0, log_1.formatRequestDetails)({
            retryOfRequestLogID,
            method: options.method,
            url,
            options,
            headers: req.headers,
        }));
        if (options.signal?.aborted) {
            throw new Errors.APIUserAbortError();
        }
        const controller = new AbortController();
        const response = await this.fetchWithTimeout(url, req, timeout, controller).catch(errors_1.castToError);
        const headersTime = Date.now();
        if (response instanceof globalThis.Error) {
            const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;
            if (options.signal?.aborted) {
                throw new Errors.APIUserAbortError();
            }
            // detect native connection timeout errors
            // deno throws "TypeError: error sending request for url (https://example/): client error (Connect): tcp connect error: Operation timed out (os error 60): Operation timed out (os error 60)"
            // undici throws "TypeError: fetch failed" with cause "ConnectTimeoutError: Connect Timeout Error (attempted address: example:443, timeout: 1ms)"
            // others do not provide enough information to distinguish timeouts from other connection errors
            const isTimeout = (0, errors_1.isAbortError)(response) ||
                /timed? ?out/i.test(String(response) + ('cause' in response ? String(response.cause) : ''));
            if (retriesRemaining) {
                (0, log_1.loggerFor)(this).info(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - ${retryMessage}`);
                (0, log_1.loggerFor)(this).debug(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (${retryMessage})`, (0, log_1.formatRequestDetails)({
                    retryOfRequestLogID,
                    url,
                    durationMs: headersTime - startTime,
                    message: response.message,
                }));
                return this.retryRequest(options, retriesRemaining, retryOfRequestLogID ?? requestLogID);
            }
            (0, log_1.loggerFor)(this).info(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - error; no more retries left`);
            (0, log_1.loggerFor)(this).debug(`[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (error; no more retries left)`, (0, log_1.formatRequestDetails)({
                retryOfRequestLogID,
                url,
                durationMs: headersTime - startTime,
                message: response.message,
            }));
            if (isTimeout) {
                throw new Errors.APIConnectionTimeoutError();
            }
            throw new Errors.APIConnectionError({ cause: response });
        }
        const responseInfo = `[${requestLogID}${retryLogStr}] ${req.method} ${url} ${response.ok ? 'succeeded' : 'failed'} with status ${response.status} in ${headersTime - startTime}ms`;
        if (!response.ok) {
            const shouldRetry = await this.shouldRetry(response);
            if (retriesRemaining && shouldRetry) {
                const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;
                // We don't need the body of this response.
                await Shims.CancelReadableStream(response.body);
                (0, log_1.loggerFor)(this).info(`${responseInfo} - ${retryMessage}`);
                (0, log_1.loggerFor)(this).debug(`[${requestLogID}] response error (${retryMessage})`, (0, log_1.formatRequestDetails)({
                    retryOfRequestLogID,
                    url: response.url,
                    status: response.status,
                    headers: response.headers,
                    durationMs: headersTime - startTime,
                }));
                return this.retryRequest(options, retriesRemaining, retryOfRequestLogID ?? requestLogID, response.headers);
            }
            const retryMessage = shouldRetry ? `error; no more retries left` : `error; not retryable`;
            (0, log_1.loggerFor)(this).info(`${responseInfo} - ${retryMessage}`);
            const errText = await response.text().catch((err) => (0, errors_1.castToError)(err).message);
            const errJSON = (0, values_1.safeJSON)(errText);
            const errMessage = errJSON ? undefined : errText;
            (0, log_1.loggerFor)(this).debug(`[${requestLogID}] response error (${retryMessage})`, (0, log_1.formatRequestDetails)({
                retryOfRequestLogID,
                url: response.url,
                status: response.status,
                headers: response.headers,
                message: errMessage,
                durationMs: Date.now() - startTime,
            }));
            const err = this.makeStatusError(response.status, errJSON, errMessage, response.headers);
            throw err;
        }
        (0, log_1.loggerFor)(this).info(responseInfo);
        (0, log_1.loggerFor)(this).debug(`[${requestLogID}] response start`, (0, log_1.formatRequestDetails)({
            retryOfRequestLogID,
            url: response.url,
            status: response.status,
            headers: response.headers,
            durationMs: headersTime - startTime,
        }));
        return { response, options, controller, requestLogID, retryOfRequestLogID, startTime };
    }
    getAPIList(path, Page, opts) {
        return this.requestAPIList(Page, { method: 'get', path, ...opts });
    }
    requestAPIList(Page, options) {
        const request = this.makeRequest(options, null, undefined);
        return new Pagination.PagePromise(this, request, Page);
    }
    async fetchWithTimeout(url, init, ms, controller) {
        const { signal, method, ...options } = init || {};
        if (signal)
            signal.addEventListener('abort', () => controller.abort());
        const timeout = setTimeout(() => controller.abort(), ms);
        const isReadableBody = (globalThis.ReadableStream && options.body instanceof globalThis.ReadableStream) ||
            (typeof options.body === 'object' && options.body !== null && Symbol.asyncIterator in options.body);
        const fetchOptions = {
            signal: controller.signal,
            ...(isReadableBody ? { duplex: 'half' } : {}),
            method: 'GET',
            ...options,
        };
        if (method) {
            // Custom methods like 'patch' need to be uppercased
            // See https://github.com/nodejs/undici/issues/2294
            fetchOptions.method = method.toUpperCase();
        }
        try {
            // use undefined this binding; fetch errors if bound to something else in browser/cloudflare
            return await this.fetch.call(undefined, url, fetchOptions);
        }
        finally {
            clearTimeout(timeout);
        }
    }
    async shouldRetry(response) {
        // Note this is not a standard header.
        const shouldRetryHeader = response.headers.get('x-should-retry');
        // If the server explicitly says whether or not to retry, obey.
        if (shouldRetryHeader === 'true')
            return true;
        if (shouldRetryHeader === 'false')
            return false;
        // Retry on request timeouts.
        if (response.status === 408)
            return true;
        // Retry on lock timeouts.
        if (response.status === 409)
            return true;
        // Retry on rate limits.
        if (response.status === 429)
            return true;
        // Retry internal errors.
        if (response.status >= 500)
            return true;
        return false;
    }
    async retryRequest(options, retriesRemaining, requestLogID, responseHeaders) {
        let timeoutMillis;
        // Note the `retry-after-ms` header may not be standard, but is a good idea and we'd like proactive support for it.
        const retryAfterMillisHeader = responseHeaders?.get('retry-after-ms');
        if (retryAfterMillisHeader) {
            const timeoutMs = parseFloat(retryAfterMillisHeader);
            if (!Number.isNaN(timeoutMs)) {
                timeoutMillis = timeoutMs;
            }
        }
        // About the Retry-After header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
        const retryAfterHeader = responseHeaders?.get('retry-after');
        if (retryAfterHeader && !timeoutMillis) {
            const timeoutSeconds = parseFloat(retryAfterHeader);
            if (!Number.isNaN(timeoutSeconds)) {
                timeoutMillis = timeoutSeconds * 1000;
            }
            else {
                timeoutMillis = Date.parse(retryAfterHeader) - Date.now();
            }
        }
        // If the API asks us to wait a certain amount of time (and it's a reasonable amount),
        // just do what it says, but otherwise calculate a default
        if (!(timeoutMillis && 0 <= timeoutMillis && timeoutMillis < 60 * 1000)) {
            const maxRetries = options.maxRetries ?? this.maxRetries;
            timeoutMillis = this.calculateDefaultRetryTimeoutMillis(retriesRemaining, maxRetries);
        }
        await (0, sleep_1.sleep)(timeoutMillis);
        return this.makeRequest(options, retriesRemaining - 1, requestLogID);
    }
    calculateDefaultRetryTimeoutMillis(retriesRemaining, maxRetries) {
        const initialRetryDelay = 0.5;
        const maxRetryDelay = 8.0;
        const numRetries = maxRetries - retriesRemaining;
        // Apply exponential backoff, but not more than the max.
        const sleepSeconds = Math.min(initialRetryDelay * Math.pow(2, numRetries), maxRetryDelay);
        // Apply some jitter, take up to at most 25 percent of the retry time.
        const jitter = 1 - Math.random() * 0.25;
        return sleepSeconds * jitter * 1000;
    }
    async buildRequest(inputOptions, { retryCount = 0 } = {}) {
        const options = { ...inputOptions };
        const { method, path, query, defaultBaseURL } = options;
        const url = this.buildURL(path, query, defaultBaseURL);
        if ('timeout' in options)
            (0, values_1.validatePositiveInteger)('timeout', options.timeout);
        options.timeout = options.timeout ?? this.timeout;
        const { bodyHeaders, body } = this.buildBody({ options });
        const reqHeaders = await this.buildHeaders({ options: inputOptions, method, bodyHeaders, retryCount });
        const req = {
            method,
            headers: reqHeaders,
            ...(options.signal && { signal: options.signal }),
            ...(globalThis.ReadableStream &&
                body instanceof globalThis.ReadableStream && { duplex: 'half' }),
            ...(body && { body }),
            ...(this.fetchOptions ?? {}),
            ...(options.fetchOptions ?? {}),
        };
        return { req, url, timeout: options.timeout };
    }
    async buildHeaders({ options, method, bodyHeaders, retryCount, }) {
        let idempotencyHeaders = {};
        if (this.idempotencyHeader && method !== 'get') {
            if (!options.idempotencyKey)
                options.idempotencyKey = this.defaultIdempotencyKey();
            idempotencyHeaders[this.idempotencyHeader] = options.idempotencyKey;
        }
        const headers = (0, headers_1.buildHeaders)([
            idempotencyHeaders,
            {
                Accept: 'application/json',
                'User-Agent': this.getUserAgent(),
                'X-Stainless-Retry-Count': String(retryCount),
                ...(options.timeout ? { 'X-Stainless-Timeout': String(Math.trunc(options.timeout / 1000)) } : {}),
                ...(0, detect_platform_1.getPlatformHeaders)(),
            },
            await this.authHeaders(options),
            this._options.defaultHeaders,
            bodyHeaders,
            options.headers,
        ]);
        this.validateHeaders(headers);
        return headers.values;
    }
    buildBody({ options: { body, headers: rawHeaders } }) {
        if (!body) {
            return { bodyHeaders: undefined, body: undefined };
        }
        const headers = (0, headers_1.buildHeaders)([rawHeaders]);
        if (
        // Pass raw type verbatim
        ArrayBuffer.isView(body) ||
            body instanceof ArrayBuffer ||
            body instanceof DataView ||
            (typeof body === 'string' &&
                // Preserve legacy string encoding behavior for now
                headers.values.has('content-type')) ||
            // `Blob` is superset of `File`
            (globalThis.Blob && body instanceof globalThis.Blob) ||
            // `FormData` -> `multipart/form-data`
            body instanceof FormData ||
            // `URLSearchParams` -> `application/x-www-form-urlencoded`
            body instanceof URLSearchParams ||
            // Send chunked stream (each chunk has own `length`)
            (globalThis.ReadableStream && body instanceof globalThis.ReadableStream)) {
            return { bodyHeaders: undefined, body: body };
        }
        else if (typeof body === 'object' &&
            (Symbol.asyncIterator in body ||
                (Symbol.iterator in body && 'next' in body && typeof body.next === 'function'))) {
            return { bodyHeaders: undefined, body: Shims.ReadableStreamFrom(body) };
        }
        else {
            return tslib_1.__classPrivateFieldGet(this, _ModernTreasury_encoder, "f").call(this, { body, headers });
        }
    }
}
exports.ModernTreasury = ModernTreasury;
_a = ModernTreasury, _ModernTreasury_encoder = new WeakMap(), _ModernTreasury_instances = new WeakSet(), _ModernTreasury_baseURLOverridden = function _ModernTreasury_baseURLOverridden() {
    return this.baseURL !== 'https://app.moderntreasury.com';
};
ModernTreasury.ModernTreasury = _a;
ModernTreasury.DEFAULT_TIMEOUT = 60000; // 1 minute
ModernTreasury.ModernTreasuryError = Errors.ModernTreasuryError;
ModernTreasury.APIError = Errors.APIError;
ModernTreasury.APIConnectionError = Errors.APIConnectionError;
ModernTreasury.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
ModernTreasury.APIUserAbortError = Errors.APIUserAbortError;
ModernTreasury.NotFoundError = Errors.NotFoundError;
ModernTreasury.ConflictError = Errors.ConflictError;
ModernTreasury.RateLimitError = Errors.RateLimitError;
ModernTreasury.BadRequestError = Errors.BadRequestError;
ModernTreasury.AuthenticationError = Errors.AuthenticationError;
ModernTreasury.InternalServerError = Errors.InternalServerError;
ModernTreasury.PermissionDeniedError = Errors.PermissionDeniedError;
ModernTreasury.UnprocessableEntityError = Errors.UnprocessableEntityError;
ModernTreasury.toFile = Uploads.toFile;
ModernTreasury.Connections = connections_1.Connections;
ModernTreasury.Counterparties = counterparties_1.Counterparties;
ModernTreasury.Events = events_1.Events;
ModernTreasury.ExpectedPayments = expected_payments_1.ExpectedPayments;
ModernTreasury.ExternalAccounts = external_accounts_1.ExternalAccounts;
ModernTreasury.IncomingPaymentDetails = incoming_payment_details_1.IncomingPaymentDetails;
ModernTreasury.Invoices = invoices_1.Invoices;
ModernTreasury.Documents = documents_1.Documents;
ModernTreasury.AccountCollectionFlows = account_collection_flows_1.AccountCollectionFlows;
ModernTreasury.AccountDetails = account_details_1.AccountDetails;
ModernTreasury.RoutingDetails = routing_details_1.RoutingDetails;
ModernTreasury.InternalAccounts = internal_accounts_1.InternalAccounts;
ModernTreasury.Ledgers = ledgers_1.Ledgers;
ModernTreasury.LedgerAccountCategories = ledger_account_categories_1.LedgerAccountCategories;
ModernTreasury.LedgerAccounts = ledger_accounts_1.LedgerAccounts;
ModernTreasury.LedgerAccountBalanceMonitors = ledger_account_balance_monitors_1.LedgerAccountBalanceMonitors;
ModernTreasury.LedgerAccountStatements = ledger_account_statements_1.LedgerAccountStatements;
ModernTreasury.LedgerEntries = ledger_entries_1.LedgerEntries;
ModernTreasury.LedgerTransactions = ledger_transactions_1.LedgerTransactions;
ModernTreasury.LineItems = line_items_1.LineItems;
ModernTreasury.PaymentFlows = payment_flows_1.PaymentFlows;
ModernTreasury.PaymentOrders = payment_orders_1.PaymentOrders;
ModernTreasury.PaymentReferences = payment_references_1.PaymentReferences;
ModernTreasury.Returns = returns_1.Returns;
ModernTreasury.Transactions = transactions_1.Transactions;
ModernTreasury.Validations = validations_1.Validations;
ModernTreasury.VirtualAccounts = virtual_accounts_1.VirtualAccounts;
ModernTreasury.BulkRequests = bulk_requests_1.BulkRequests;
ModernTreasury.BulkResults = bulk_results_1.BulkResults;
ModernTreasury.LedgerAccountSettlements = ledger_account_settlements_1.LedgerAccountSettlements;
ModernTreasury.ForeignExchangeQuotes = foreign_exchange_quotes_1.ForeignExchangeQuotes;
ModernTreasury.ConnectionLegalEntities = connection_legal_entities_1.ConnectionLegalEntities;
ModernTreasury.LegalEntities = legal_entities_1.LegalEntities;
ModernTreasury.LegalEntityAssociations = legal_entity_associations_1.LegalEntityAssociations;
ModernTreasury.PaymentActions = payment_actions_1.PaymentActions;
