// modern-treasury/resources/invoices/invoices.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as LineItemsAPI from "./line-items.mjs";
import { LineItems, } from "./line-items.mjs";
import { Page } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Invoices extends APIResource {
    constructor() {
        super(...arguments);
        this.lineItems = new LineItemsAPI.LineItems(this._client);
    }
    /**
     * create invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.create({
     *   counterparty_id: 'counterparty_id',
     *   due_date: '2019-12-27T18:11:19.117Z',
     *   originating_account_id: 'originating_account_id',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/invoices', { body, ...options });
    }
    /**
     * get invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/invoices/${id}`, options);
    }
    /**
     * update invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/invoices/${id}`, { body, ...options });
    }
    /**
     * list invoices
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const invoice of client.invoices.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/invoices', (Page), { query, ...options });
    }
    /**
     * Add a payment order to an invoice.
     *
     * @example
     * ```ts
     * await client.invoices.addPaymentOrder('payment_order_id', {
     *   id: 'id',
     * });
     * ```
     */
    addPaymentOrder(paymentOrderID, params, options) {
        const { id } = params;
        return this._client.put(path `/api/invoices/${id}/payment_orders/${paymentOrderID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
Invoices.LineItems = LineItems;
