// modern-treasury/resources/invoices/invoices.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Invoices = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const LineItemsAPI = tslib_1.__importStar(require("./line-items.js"));
const line_items_1 = require("./line-items.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Invoices extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.lineItems = new LineItemsAPI.LineItems(this._client);
    }
    /**
     * create invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.create({
     *   counterparty_id: 'counterparty_id',
     *   due_date: '2019-12-27T18:11:19.117Z',
     *   originating_account_id: 'originating_account_id',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/invoices', { body, ...options });
    }
    /**
     * get invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/invoices/${id}`, options);
    }
    /**
     * update invoice
     *
     * @example
     * ```ts
     * const invoice = await client.invoices.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/invoices/${id}`, { body, ...options });
    }
    /**
     * list invoices
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const invoice of client.invoices.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/invoices', (pagination_1.Page), { query, ...options });
    }
    /**
     * Add a payment order to an invoice.
     *
     * @example
     * ```ts
     * await client.invoices.addPaymentOrder('payment_order_id', {
     *   id: 'id',
     * });
     * ```
     */
    addPaymentOrder(paymentOrderID, params, options) {
        const { id } = params;
        return this._client.put((0, path_1.path) `/api/invoices/${id}/payment_orders/${paymentOrderID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Invoices = Invoices;
Invoices.LineItems = line_items_1.LineItems;
