// modern-treasury/resources/invoices/line-items.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineItems = void 0;
const resource_1 = require("../../core/resource.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class LineItems extends resource_1.APIResource {
    /**
     * create invoice_line_item
     *
     * @example
     * ```ts
     * const invoiceLineItem =
     *   await client.invoices.lineItems.create('invoice_id', {
     *     name: 'name',
     *     unit_amount: 0,
     *   });
     * ```
     */
    create(invoiceID, body, options) {
        return this._client.post((0, path_1.path) `/api/invoices/${invoiceID}/invoice_line_items`, { body, ...options });
    }
    /**
     * get invoice_line_item
     *
     * @example
     * ```ts
     * const invoiceLineItem =
     *   await client.invoices.lineItems.retrieve('id', {
     *     invoice_id: 'invoice_id',
     *   });
     * ```
     */
    retrieve(id, params, options) {
        const { invoice_id } = params;
        return this._client.get((0, path_1.path) `/api/invoices/${invoice_id}/invoice_line_items/${id}`, options);
    }
    /**
     * update invoice_line_item
     *
     * @example
     * ```ts
     * const invoiceLineItem =
     *   await client.invoices.lineItems.update('id', {
     *     invoice_id: 'invoice_id',
     *   });
     * ```
     */
    update(id, params, options) {
        const { invoice_id, ...body } = params;
        return this._client.patch((0, path_1.path) `/api/invoices/${invoice_id}/invoice_line_items/${id}`, {
            body,
            ...options,
        });
    }
    /**
     * list invoice_line_items
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const invoiceLineItem of client.invoices.lineItems.list(
     *   'invoice_id',
     * )) {
     *   // ...
     * }
     * ```
     */
    list(invoiceID, query = {}, options) {
        return this._client.getAPIList((0, path_1.path) `/api/invoices/${invoiceID}/invoice_line_items`, (pagination_1.Page), { query, ...options });
    }
    /**
     * delete invoice_line_item
     *
     * @example
     * ```ts
     * const invoiceLineItem =
     *   await client.invoices.lineItems.delete('id', {
     *     invoice_id: 'invoice_id',
     *   });
     * ```
     */
    delete(id, params, options) {
        const { invoice_id } = params;
        return this._client.delete((0, path_1.path) `/api/invoices/${invoice_id}/invoice_line_items/${id}`, options);
    }
}
exports.LineItems = LineItems;
