// modern-treasury/resources/transactions/line-items.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineItems = void 0;
const resource_1 = require("../../core/resource.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class LineItems extends resource_1.APIResource {
    /**
     * create transaction line items
     *
     * @example
     * ```ts
     * const transactionLineItem =
     *   await client.transactions.lineItems.create({
     *     amount: 0,
     *     expected_payment_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     transaction_id: '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/transaction_line_items', { body, ...options });
    }
    /**
     * get transaction line item
     *
     * @example
     * ```ts
     * const transactionLineItem =
     *   await client.transactions.lineItems.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/transaction_line_items/${id}`, options);
    }
    /**
     * list transaction_line_items
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const transactionLineItem of client.transactions.lineItems.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/transaction_line_items', (pagination_1.Page), {
            query,
            ...options,
        });
    }
    /**
     * delete transaction line item
     *
     * @example
     * ```ts
     * await client.transactions.lineItems.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/transaction_line_items/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.LineItems = LineItems;
