// modern-treasury/resources/transactions/line-items.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { Page } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class LineItems extends APIResource {
    /**
     * create transaction line items
     *
     * @example
     * ```ts
     * const transactionLineItem =
     *   await client.transactions.lineItems.create({
     *     amount: 0,
     *     expected_payment_id:
     *       '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *     transaction_id: '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/transaction_line_items', { body, ...options });
    }
    /**
     * get transaction line item
     *
     * @example
     * ```ts
     * const transactionLineItem =
     *   await client.transactions.lineItems.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/transaction_line_items/${id}`, options);
    }
    /**
     * list transaction_line_items
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const transactionLineItem of client.transactions.lineItems.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/transaction_line_items', (Page), {
            query,
            ...options,
        });
    }
    /**
     * delete transaction line item
     *
     * @example
     * ```ts
     * await client.transactions.lineItems.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/transaction_line_items/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
