// modern-treasury/resources/transactions/transactions.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as LineItemsAPI from "./line-items.mjs";
import { LineItems, } from "./line-items.mjs";
import { Page } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Transactions extends APIResource {
    constructor() {
        super(...arguments);
        this.lineItems = new LineItemsAPI.LineItems(this._client);
    }
    /**
     * create transaction
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.create({
     *   amount: 0,
     *   as_of_date: '2019-12-27',
     *   direction: 'direction',
     *   internal_account_id:
     *     '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   vendor_code: 'vendor_code',
     *   vendor_code_type: 'vendor_code_type',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/transactions', { body, ...options });
    }
    /**
     * Get details on a single transaction.
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/transactions/${id}`, options);
    }
    /**
     * Update a single transaction.
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/transactions/${id}`, { body, ...options });
    }
    /**
     * Get a list of all transactions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const transaction of client.transactions.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/transactions', (Page), { query, ...options });
    }
    /**
     * delete transaction
     *
     * @example
     * ```ts
     * await client.transactions.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/api/transactions/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
Transactions.LineItems = LineItems;
