// modern-treasury/resources/transactions/transactions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transactions = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const LineItemsAPI = tslib_1.__importStar(require("./line-items.js"));
const line_items_1 = require("./line-items.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Transactions extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.lineItems = new LineItemsAPI.LineItems(this._client);
    }
    /**
     * create transaction
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.create({
     *   amount: 0,
     *   as_of_date: '2019-12-27',
     *   direction: 'direction',
     *   internal_account_id:
     *     '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   vendor_code: 'vendor_code',
     *   vendor_code_type: 'vendor_code_type',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/transactions', { body, ...options });
    }
    /**
     * Get details on a single transaction.
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.retrieve(
     *   'id',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/transactions/${id}`, options);
    }
    /**
     * Update a single transaction.
     *
     * @example
     * ```ts
     * const transaction = await client.transactions.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/transactions/${id}`, { body, ...options });
    }
    /**
     * Get a list of all transactions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const transaction of client.transactions.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/transactions', (pagination_1.Page), { query, ...options });
    }
    /**
     * delete transaction
     *
     * @example
     * ```ts
     * await client.transactions.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/api/transactions/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Transactions = Transactions;
Transactions.LineItems = line_items_1.LineItems;
