// modern-treasury/resources/ledger-transactions/ledger-transactions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerTransactions = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const VersionsAPI = tslib_1.__importStar(require("./versions.js"));
const versions_1 = require("./versions.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class LedgerTransactions extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.versions = new VersionsAPI.Versions(this._client);
    }
    /**
     * Create a ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.create({
     *     ledger_entries: [
     *       {
     *         amount: 0,
     *         direction: 'credit',
     *         ledger_account_id:
     *           '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *       },
     *     ],
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_transactions', { body, ...options });
    }
    /**
     * Get details on a single ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/ledger_transactions/${id}`, options);
    }
    /**
     * Update the details of a ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/api/ledger_transactions/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger transactions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerTransaction of client.ledgerTransactions.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_transactions', (pagination_1.Page), {
            query,
            ...options,
        });
    }
    /**
     * Create a ledger transaction that partially posts another ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.createPartialPost('id', {
     *     posted_ledger_entries: [
     *       {
     *         amount: 0,
     *         direction: 'credit',
     *         ledger_account_id:
     *           '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *       },
     *     ],
     *   });
     * ```
     */
    createPartialPost(id, body, options) {
        return this._client.post((0, path_1.path) `/api/ledger_transactions/${id}/partial_post`, { body, ...options });
    }
    /**
     * Create a ledger transaction reversal.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.createReversal('id');
     * ```
     */
    createReversal(id, body = {}, options) {
        return this._client.post((0, path_1.path) `/api/ledger_transactions/${id}/reversal`, { body, ...options });
    }
}
exports.LedgerTransactions = LedgerTransactions;
LedgerTransactions.Versions = versions_1.Versions;
