// modern-treasury/resources/ledger-transactions/ledger-transactions.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as VersionsAPI from "./versions.mjs";
import { Versions, } from "./versions.mjs";
import { Page } from "../../core/pagination.mjs";
import { path } from "../../internal/utils/path.mjs";
export class LedgerTransactions extends APIResource {
    constructor() {
        super(...arguments);
        this.versions = new VersionsAPI.Versions(this._client);
    }
    /**
     * Create a ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.create({
     *     ledger_entries: [
     *       {
     *         amount: 0,
     *         direction: 'credit',
     *         ledger_account_id:
     *           '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *       },
     *     ],
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/ledger_transactions', { body, ...options });
    }
    /**
     * Get details on a single ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/api/ledger_transactions/${id}`, options);
    }
    /**
     * Update the details of a ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch(path `/api/ledger_transactions/${id}`, { body, ...options });
    }
    /**
     * Get a list of ledger transactions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ledgerTransaction of client.ledgerTransactions.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/ledger_transactions', (Page), {
            query,
            ...options,
        });
    }
    /**
     * Create a ledger transaction that partially posts another ledger transaction.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.createPartialPost('id', {
     *     posted_ledger_entries: [
     *       {
     *         amount: 0,
     *         direction: 'credit',
     *         ledger_account_id:
     *           '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *       },
     *     ],
     *   });
     * ```
     */
    createPartialPost(id, body, options) {
        return this._client.post(path `/api/ledger_transactions/${id}/partial_post`, { body, ...options });
    }
    /**
     * Create a ledger transaction reversal.
     *
     * @example
     * ```ts
     * const ledgerTransaction =
     *   await client.ledgerTransactions.createReversal('id');
     * ```
     */
    createReversal(id, body = {}, options) {
        return this._client.post(path `/api/ledger_transactions/${id}/reversal`, { body, ...options });
    }
}
LedgerTransactions.Versions = Versions;
